tuning/trainercontroller/callback.py

```diff
@@ -0,0 +1,353 @@
+                    self.metrics_on_event[event_name].append(obj)
+                else:
+                    raise KeyError(f"Event name ({event_name}) is not valid in metric {metric_name}")
+
+        # Check if there any operations listed in the configuration
+        if OPERATIONS_KEY in self.trainer_controller_config:
+            # Operation handler validation and registration is performed here.
+            for operation_name, operation_config in self.trainer_controller_config[OPERATIONS_KEY].items():
+                # Get the operation class name from the config section.
+                operation_handler_name = next(iter(operation_config.keys()))
+                # Get the handler class arguments using the operation class name.
+                operation_args = operation_config[operation_handler_name]
+                if operation_args is None:
+                    operation_args = {}
+                # Operation handler instance is created here.
+                operation = \
+                    self.operation_handlers[operation_handler_name](**operation_args, **kwargs)
+                # Add operation action instances.
+                for action_name in operation.get_actions():
+                    self.operation_actions[operation_name+"."+action_name] = OperationAction(instance=operation, action=action_name)
+
+        # Initialize controllers with respect to events.
+        if CONTROLLERS_KEY in self.trainer_controller_config:
+            for controller in self.trainer_controller_config[CONTROLLERS_KEY]:
+                for event_name in controller[CONTROLLER_TRIGGERS_KEY]:
+                    if event_name not in self.valid_events:
+                        raise KeyError(f"Controller {controller[CONTROLLER_NAME_KEY]} has an invalid event ({event_name})")
+                    # Generates the byte-code for the rule from trainer configuration
+                    if not self._validate_rule(controller[CONTROLLER_RULE_KEY]):
+                        raise ValueError(f"Rule for control {controller[CONTROLLER_NAME_KEY]} is invalid")
+                    control = Control(name=controller[CONTROLLER_NAME_KEY], rule = compile(controller[CONTROLLER_RULE_KEY], '', 'eval'), operation_actions = [])
+                    for control_operation_name in controller[CONTROLLER_OPERATIONS_KEY]:
+                        if control_operation_name not in self.operation_actions:
+                            raise KeyError(f"Invalid operation {control_operation_name} for control {controller[CONTROLLER_NAME_KEY]}")
+                        control.operation_actions.append(self.operation_actions[control_operation_name])
+                    if event_name not in self.control_actions_on_event:
+                        self.control_actions_on_event[event_name] = []
+                    self.control_actions_on_event[event_name].append(control)
+                        
+        self._actions_on_event(event_name='on_init_end', **kwargs)
```

Comment:
Same question here - in what situation is it useful invoke an action on init end if the focus of this is metric based and dependent on things that are being produced as training progresses?

Reply:
@alex-jw-brooks We could perform actions like validate configs and  if validation fails `stop the training` before it starts or Notify user on training started or is about start. It has some uses and hence we have left it to cover any such future use cases.